config/initializers/rack_attack.rb

```diff
@@ -68,12 +84,42 @@ def remote_ip
     # Throttle sign in attempts by IP address
     #
     # Key: "rack::attack:#{Time.now.to_i/:period}:logins/ip:#{req.remote_ip}"
-    throttle(
-      'logins/ip',
-      limit: Figaro.env.logins_per_ip_limit.to_i,
-      period: Figaro.env.logins_per_ip_period.to_i.seconds
-    ) do |req|
-      req.remote_ip if req.path == '/' && req.post?
+    if Figaro.env.logins_per_ip_track_only_mode == 'true'
+      track(
+        'logins/ip',
+        limit: Figaro.env.logins_per_ip_limit.to_i,
+        period: Figaro.env.logins_per_ip_period.to_i
+      ) do |req|
+        req.remote_ip if req.path == '/' && req.post?
+      end
+    else
+      throttle(
+        'logins/ip',
+        limit: Figaro.env.logins_per_ip_limit.to_i,
+        period: Figaro.env.logins_per_ip_period.to_i
+      ) do |req|
+        req.remote_ip if req.path == '/' && req.post?
+      end
+    end
+
+    # Lockout IP addresses that are attempting to sign in with the same username
+    # over and over.
+    # After maxretry requests in findtime minutes, block all requests from that IP for bantime.
+    blocklist('logins/email+ip') do |req|
+      # `filter` returns false if POST request is for the login page (but still
+      # increments the count), so requests below the limit are not blocked until
+      # they hit the limit. At that point, `filter` will return true and block.
+      user = req.params.fetch('user', {})
+      email = user['email']
+      email_and_ip = "#{email}-#{req.remote_ip}"
```

Comment:
We may want to pass the email through the PII fingerprinter before writing it here so that plaintext emails aren't going into redis

Reply:
Additionally, is there a way we could break or ignore and let the other limits take over if the request does not include an email param?

Reply:
Good catch on the fingerprinting. As for the blank email, I thought about it but couldn't think of anything bad happening. Most legitimate users won't not be submitting a blank email because most browsers will prevent you from doing that, even with JS disabled. If someone is deliberately passing in a blank email somehow, they should be blocked. Am I missing anything?

Reply:
Added fingerprint in https://github.com/18F/identity-idp/pull/1902/commits/ed6d629bbdbe23c9b44639d7ee989b13d567c0cc. PTAL